File: NeetCode/Trees/House_Robber_III/house__robber__i_i_i.py
from typing import Optional, List
import heapq
from collections import defaultdict

from common.treenode import TreeNode, deserialize, printTree, null


class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        # return (canRob, cantRob)
        def dfs(node):
            if not node:
                return (0, 0)

            left = dfs(node.left)
            right = dfs(node.right)

            canRob = node.val + left[1] + right[1]
            dontRob = max(left) + max(right)

            return (canRob, dontRob)

        return max(dfs(root))


if __name__ == "__main__":
    sol = Solution()
    root = [4, 1, null, 2, null, 3]
    root = deserialize(root)

    print(sol.rob(root))
    print("Running Solution...")

File: NeetCode/Trees/House_Robber_III/House_Robber_III.cpp
// House_Robber_III.cpp
//
// Compile locally with:
// clang++ House_Robber_III.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution
{
public:
    int rob(TreeNode *root)
    {
        pair<int, int> results = dfs(root);
        return max({results.first, results.second});
    }
    pair<int, int> dfs(TreeNode *node)
    {
        if (node == nullptr)
        {
            return {0, 0};
        }

        pair<int, int> left = dfs(node->left);
        pair<int, int> right = dfs(node->right);

        int canRob = node->val + left.second + right.second;
        int cantRob = max({left.first, left.second}) + max({right.first, right.second});
        return {canRob, cantRob};
    }
};

int main()
{
    cout << "Running House_Robber_III.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}
